### ISSUE TITLE
[DOC]/[ENH]: Fix type annotation mismatches and add input validation in TslibDataModule

### ISSUE THREAD

**Issue by phantom-712** [jdb78/pytorch-forecasting#2153]
# DOC/ENH: Fix type annotation mismatches and add input validation in `TslibDataModule`

## Summary

`TslibDataModule` in `pytorch_forecasting/data/_tslib_data_module.py` has several type annotation mismatches between method signatures and their docstrings, a malformed property docstring, and no input validation on critical `__init__` parameters.

These issues affect IDE support, static type checking (mypy), auto-generated API docs, and user experience when misconfigured inputs cause cryptic downstream errors instead of clear, early `ValueError`s.

This issue proposes fixing all of these in a single, focused PR touching only
`_tslib_data_module.py`.

---

### 1. `_preprocess_data` - parameter annotation says `torch.Tensor`, but receives `int`

The method signature declares:

```python
def _preprocess_data(self, idx: torch.Tensor) -> dict[str, Any]:
```

But every call site in `_create_windows` extracts a plain Python `int` before
passing it in:

```python
series_idx = idx.item() if isinstance(idx, torch.Tensor) else idx
# series_idx is now a plain int

processed_data = self.data_module._preprocess_data(series_idx)  # int, not Tensor
```

This is a real type mismatch, not just a documentation issue. mypy will flag any
caller that passes an `int`, and any contributor subclassing `TslibDataModule` will
be misled about what the method actually expects.

**Fix:**

```python
def _preprocess_data(self, idx: int) -> dict[str, Any]:
    """
    ...
    Parameters
    ----------
    idx : int          # ← was torch.Tensor
        The index of the time series data to be processed.
    ...
    """
```

---

### 2. `_preprocess_data` - return type annotation says `list[dict]`, body returns `dict`

The current annotation:

```python
def _preprocess_data(self, idx: torch.Tensor) -> list[dict[str, Any]]:
```

The actual return statement:

```python
res = {
    "features": {...},
    "target": target,
    "static": series["st"],
    ...
}
return res  # a single dict, not a list
```

**Fix:**

```python
def _preprocess_data(self, idx: int) -> dict[str, Any]:
```

---

### 3. `collate_fn` - docstring contradicts its own signature

After the signature was recently annotated correctly as:

```python
@staticmethod
def collate_fn(
    batch: list[
        tuple[
            dict[str, Any],
            torch.Tensor | list[torch.Tensor],
        ]
    ],
) -> tuple[dict[str, torch.Tensor], torch.Tensor | list[torch.Tensor]]:
```

The `Parameters` section of the docstring was never updated and still reads:

```python
"""
Parameters
----------
batch: list[tuple[dict[str, Any]]]
    The batch of data to be collated.
"""
```

This is missing the second element of the tuple entirely, meaning the signature
and docstring actively contradict each other.

**Fix - update docstring to match the signature:**

```python
"""
Parameters
----------
batch : list[tuple[dict[str, Any], torch.Tensor or list of torch.Tensor]]
    A list of (x, y) tuples where:
    - x is a dict of named feature tensors
    - y is either a single target tensor or a list of tensors for
      multi-target datasets
"""
```

---

### 4. `metadata` property - malformed docstring opening

The current docstring opens with a stray `" "` character:

```python
@property
def metadata(self) -> dict[str, Any]:
    """ "
    Compute the metadata via the `_prepare_metadata` method.
```

This corrupts auto-generated API documentation.

**Fix:**

```python
@property
def metadata(self) -> dict[str, Any]:
    """
    Metadata for the time series dataset.

    Returns
    -------
    dict[str, Any]
        Metadata for the data module. Refer to `_prepare_metadata`
        for the keys and values in the metadata dictionary.
    """
```

---

### 5. `_validate_indices` - missing `-> None` return annotation

```python
def _validate_indices(self):   # ← no return annotation
```

**Fix:**

```python
def _validate_indices(self) -> None:
```

---

### 6. No input validation on critical `__init__` parameters

`__init__` currently accepts `context_length`, `prediction_length`, `window_stride`,
and `train_val_test_split` without any validation. Invalid values propagate silently
into `_create_windows` and produce empty window lists or index errors with no
indication of what went wrong.

Example - passing `context_length=-1` currently produces no error at construction
time and silently results in no training windows being generated.

**Fix - add guard clauses in `__init__` after `super().__init__()`:**

```python
if context_length <= 0:
    raise ValueError(
        f"context_length must be a positive integer, got {context_length}."
    )

if prediction_length <= 0:
    raise ValueError(
        f"prediction_length must be a positive integer, got {prediction_length}."
    )

if window_stride <= 0:
    raise ValueError(
        f"window_stride must be a positive integer, got {window_stride}."
    )

if not (
    len(train_val_test_split) == 3
    and abs(sum(train_val_test_split) - 1.0) < 1e-6
    and all(s > 0 for s in train_val_test_split)
):
    raise ValueError(
        "train_val_test_split must be a tuple of 3 positive floats that sum "
        f"to 1.0, got {train_val_test_split}."
    )
```

---

## Scope

| | |
|---|---|
| **File** | `pytorch_forecasting/data/_tslib_data_module.py` only |
| **Lines changed** | ~35–45 (annotations, docstring fixes, 4 guard clauses) |
| **Runtime behaviour** | Unchanged except new `ValueError`s on invalid inputs |
| **Tests needed** | Existing v2 model tests (`test_timexer_v2`, `test_dlinear_v2`) should continue to pass. Optionally: small parametrized test for the new `ValueError`s. |

---

## Why This Matters

`TslibDataModule` is the core D2 layer of the v2 architecture, the first thing
every new v2 model and every new contributor interacts with. As the v2 API
stabilises and more contributors build on it, having accurate type information
and early, descriptive errors significantly reduces friction.

The `_preprocess_data` type mismatch in particular (`torch.Tensor` vs `int`) is the
kind of bug that causes real confusion when subclassing `TslibDataModule` to build
new D2 integrations, which is exactly what the ESoC 2026 foundation model projects
will require.

---

## Checklist for the PR

- [ ] `_preprocess_data` parameter annotation changed from `torch.Tensor` to `int`
- [ ] `_preprocess_data` return annotation changed from `list[dict[str, Any]]` to `dict[str, Any]`
- [ ] `collate_fn` docstring `Parameters` section updated to match existing signature
- [ ] `metadata` property docstring opening fixed
- [ ] `_validate_indices` annotated with `-> None`
- [ ] Four `ValueError` guard clauses added in `__init__`
- [ ] All existing v2 tests pass locally
- [ ] flake8 clean at `--max-line-length 100`

---

**Labels:** `good first issue` · `enhancement` · `documentation` · `module:data`

**Comment by phantom-712** [comment]
Dear maintainers, can this issue be assigned to me?
I've already read through _tslib_data_module.py in detail and have a local branch with the fixes ready. Happy to have it reviewed before merging
